include/alpaka/mem/MdSpanArray.hpp

```diff
@@ -0,0 +1,251 @@
+        {
+            return ResolveArrayAccess<T_numDims - 1u, T_dim + 1u>{}(arrayPtr[idx[T_dim]], idx);
+        }
+    };
+
+    template<uint32_t T_dim>
+    struct ResolveArrayAccess<1u, T_dim>
+    {
+        constexpr decltype(auto) operator()(auto arrayPtr, concepts::Vector auto const& idx) const
+        {
+            return arrayPtr[idx[T_dim]];
+        }
+    };
+
+    /** build C array type with compile time extents from a scalar value based on the compile time extents vector */
+    template<typename T, concepts::CVector T_Extents, uint32_t T_numDims = T_Extents::dim(), uint32_t T_dim = 0u>
+    struct CArrayType
+    {
+        using type =
+            typename CArrayType<T[T_Extents{}[T_numDims - T_dim - 1u]], T_Extents, T_numDims - 1u, T_dim + 1u>::type;
+    };
+
+    template<typename T, concepts::CVector T_Extents, uint32_t T_dim>
+    struct CArrayType<T, T_Extents, 1u, T_dim>
+    {
+        using type = T[T_Extents{}[0u]];
+    };
+
+    template<typename T_ArrayType, concepts::Alignment T_MemAlignment = Alignment<>>
+    struct MdSpanArray
+    {
+        static_assert(
+            sizeof(T_ArrayType) && false,
+            "MdSpanArray can only be used if std::is_array_v<T> is true for the given type.");
+    };
+
+    template<alpaka::concepts::CStaticArray T_ArrayType, concepts::Alignment T_MemAlignment>
+    struct MdSpanArray<T_ArrayType, T_MemAlignment>
+    {
+    private:
+        using MutArrayType = std::remove_cv_t<T_ArrayType>;
+        using ConstArrayType = std::add_const_t<MutArrayType>;
+
+    public:
+        using extentType = std::extent<T_ArrayType, std::rank_v<T_ArrayType>>;
+        using value_type = std::remove_all_extents_t<T_ArrayType>;
+        using reference = value_type&;
+        using const_reference = value_type const&;
+        using pointer = value_type*;
+        using const_pointer = value_type const*;
+        using index_type = typename extentType::value_type;
+
+        static consteval uint32_t dim()
+        {
+            return std::rank_v<T_ArrayType>;
+        }
+
+        /** return value the origin pointer is pointing to
+         *
+         * @return value at the current location
+         */
+        constexpr const_reference operator*() const
+        {
+            return *this->m_ptr;
+        }
+
+        constexpr reference operator*()
+        {
+            return *this->m_ptr;
+        }
+
+        /** get origin pointer */
+        constexpr const_pointer data() const
+        {
+            return this->m_ptr;
+        }
+
+        constexpr pointer data()
+        {
+            return this->m_ptr;
+        }
+
+        constexpr auto begin() const
+        {
+            return MdForwardIter{*this};
+        }
+
+        constexpr auto end() const
+        {
+            return MdForwardIterEnd{*this};
+        }
+
+        constexpr auto getConstMdSpan() const
+        {
+            return MdSpanArray<ConstArrayType>(*m_ptr);
```

Comment:
here is a bug I introduced already before this PR.
The Alignment is missing. I will fix this after merging this PR.